presto-main/src/main/java/io/prestosql/operator/scalar/UrlFunctions.java

```diff
@@ -128,15 +128,19 @@ public static Slice urlExtractParameter(@SqlType("varchar(x)") Slice url, @SqlTy
             return null;
         }
 
-        Slice query = slice(uri.getQuery());
         String parameter = parameterName.toStringUtf8();
-        Iterable<String> queryArgs = QUERY_SPLITTER.split(query.toStringUtf8());
+        Iterable<String> queryArgs = QUERY_SPLITTER.split(uri.getRawQuery());
 
         for (String queryArg : queryArgs) {
             Iterator<String> arg = ARG_SPLITTER.split(queryArg).iterator();
             if (arg.next().equals(parameter)) {
                 if (arg.hasNext()) {
-                    return utf8Slice(arg.next());
+                    try {
+                        return utf8Slice(URLDecoder.decode(arg.next(), UTF_8.name()));
```

Comment:
This will throw `IllegalArgumentException` on invalid input, which we should handle. We can reuse the `urlDecode` method (maybe extract an internal version that takes `String` instead of `Slice`).